File: 2024/April/4-26 1289 Minimum Falling Path Sum II/MinFallingPath.py
from typing import List

class Solution:
    def minFallingPathSum(self, grid: List[List[int]]) -> int:
        def minExclusive(row: List[int], exclude: int) -> int:
            rowMin = 1000000000
            for i, num in enumerate(row):
                if i == exclude:
                    continue
                if num < rowMin:
                    rowMin = num
            return rowMin

        previous = grid[len(grid) - 1]
        for i in range(len(grid) - 2, -1, -1):
            for j in range(len(previous)):
                grid[i][j] += minExclusive(previous, j)
            previous = grid[i]
        return minExclusive(previous, -1)
    

File: 2024/April/4-26 1289 Minimum Falling Path Sum II/MinFallingPath.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minFallingPathSum = function(grid) {
    var minExclusive = (row, exclude) => {
        let rowMin = Number.MAX_SAFE_INTEGER;
        for(let i = 0; i < row.length; i++) {
            if(i === exclude) {
                continue;
            }

            if(row[i] < rowMin) {
                rowMin = row[i];
            }
        }
        return rowMin;
    };

    let previous = grid[grid.length - 1];
    for(let i = grid.length - 2; i >= 0; i--) {
        for(let j = 0; j < previous.length; j++) {
            grid[i][j] += minExclusive(previous, j);
        }
        previous = grid[i];
    }

    return minExclusive(previous, -1);
};